### DISCUSSION TITLE
React Aria Components + React Hook Form

### DISCUSSION THREAD

**Discussion by thevipinmishra** [react-hook-form/react-hook-form#10958] [Q&A]
I'm having some issues with the FileTrigger component of the React Aria Components when I use it with `register` method. I'm unable to debug it as well.

Can someone please have a look and help?

Here's the [FileTrigger API](https://react-spectrum.adobe.com/react-aria/FileTrigger.html)

Repro - https://codesandbox.io/p/sandbox/determined-flower-cgv6nh?welcome=true

**Comment by Moshyfawn** [comment]
You can't use the `register` method with `FileTrigger` directly as the input's `onChange` event doesn't return the native Web `event` with `target` _(ref adobe/react-spectrum#4854)_.

You can try to override the `onChange` handler returned from `register` by passing an additional `onChange` prop with [`setValue`](https://react-hook-form.com/docs/useform/setvalue) and pass the `FileList` returned from `FileTrigger` `onChange` directly to the form values setter.

```jsx
const { register, setValue } = useForm()

return (
  <FileTrigger {...register("file")}  onChange={file => setValue('file', file)}>
    ...
  </FileTrigger>
)
```

Otherwise, consider converting the input into a controlled input via the [`useController`](https://react-hook-form.com/docs/usecontroller) / [`Controller`](https://react-hook-form.com/docs/usecontroller/controller) API.

**Comment by thevipinmishra** [comment]
> You can't use the `register` method with `FileTrigger` directly as the input's `onChange` event doesn't return the native Web `event` with `target` _(ref [adobe/react-spectrum#4854](https://github.com/adobe/react-spectrum/issues/4854))_.
> 
> You can try to override the `onChange` handler returned from `register` by passing an additional `onChange` prop with [`setValue`](https://react-hook-form.com/docs/useform/setvalue) and pass the `FileList` returned from `FileTrigger` `onChange` directly to the form values setter.
> 
> ```js
> const { register, setValue } = useForm()
> 
> return (
>   <FileTrigger {...register("file")}  onChange={file => setValue('file', file)}>
>     ...
>   </FileTrigger>
> )
> ```
> 
> Otherwise, consider converting the input into a controlled input via the [`useController`](https://react-hook-form.com/docs/usecontroller) / [`Controller`](https://react-hook-form.com/docs/usecontroller/controller) API.

Hi, thanks for this.

Can you please give me an example with Controller ?

**Reply by Moshyfawn** [reply]
Here's a very basic setup. The rest of it will depend on your specific use-case

```jsx
<Controller
  control={control}
  name="file"
  render={({ field }) => (
    <FileTrigger
      {...field}
      onChange={(files) => {
        createPreview(files[0]); // INFO: this can be a file upload API call that returns an image preview or something like a `FileReader` Web API call that returns `base64`, for example
        field.onChange(files[0])
      }}
    >
      <Button>Select a file</Button>
    </FileTrigger>
  )}
/>
```

Here's a basic example `createPreview` implementation

```js
const createPreview = (file) => {
  const reader = new FileReader()
  reader.readAsDataURL(file)
  reader.onload = () => {
    setFilePreview(reader.result); // INFO: you can setup a React state to store the preview `string`
  }
  return filePreview
}
```

Here's [a quick and dirty codesandbox example](https://codesandbox.io/p/sandbox/frosty-violet-mw2w3w?file=%2Fsrc%2FApp.js%3A37%2C13)
